### PR TITLE
Add targets

### PR DESCRIPTION
**Pull request by ivanfelfo**
created Topic model & table.
created Target model, controller & table

Signed in user can create and destroy targets

### CODE REVIEW

File: db/migrate/20210503145415_create_targets.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class CreateTargets < ActiveRecord::Migration[6.1]
+  def change
+    create_table :targets do |t|
+      t.integer :topic_id
+      t.integer :user_id
+      t.string :title
+      t.decimal :radius
+      t.decimal :lon
+      t.decimal :lat
```
Comment (juanmferreira93) [inline db/migrate/20210503145415_create_targets.rb:9]: maybe rename this as `latitude` and `longitude`

File: spec/factories/targets_factories.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+  factory :target do
+    title { Faker::Lorem.sentence }
+    radius { Faker::Number.decimal(2, 3) }
+    lat { Faker::Number.decimal(2, 3) }
+    lon { Faker::Number.decimal(2, 3) }
+    user
+    topic
```
Comment (juanmferreira93) [inline spec/factories/targets_factories.rb:8]: here you need to add the association with User and Topic models.
Example: `association :user, factory: :user`

File: spec/requests/targets/targets_create_spec.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  let(:topic) { create :topic }
+  subject { post v1_targets_path, params: params, as: :json }
+
+  context 'when user is logged in' do
+    sign_in(:user)
+    let(:params) do
+      { target: { title: 'hola', lat: 1, lon: 1, radius: 2, topic_id: topic[:id] } }
+    end
+
+    it 'returns http success' do
```
Comment (juanmferreira93) [inline spec/requests/targets/targets_create_spec.rb:16]: could you spec that the amount of targets have change, check that the target is created with the correct values, etc. Think about all possible cases and things that are valuable to test

File: app/models/user.rb [hunk 4]
```diff
@@ -6,6 +6,8 @@ class User < ApplicationRecord
          :recoverable, :rememberable, :validatable, :confirmable
   include DeviseTokenAuth::Concerns::User
 
+  has_many :targets, dependent: :destroy
```
Comment (juanmferreira93) [inline app/models/user.rb:9]: this need to be tested

File: app/models/target.rb [hunk 5]
```diff
@@ -0,0 +1,4 @@
+class Target < ApplicationRecord
```
Comment (juanmferreira93) [inline app/models/target.rb:1]: specs needed here

File: app/models/topic.rb [hunk 6]
```diff
@@ -0,0 +1,3 @@
+class Topic < ApplicationRecord
```
Comment (juanmferreira93) [inline app/models/topic.rb:1]: specs needed here

File: spec/requests/targets/targets_create_spec.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'POST v1/targets', type: :request do
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:5]:
```suggestion
describe 'POST v1/targets', type: :request do
```

File: spec/factories/targets_factories.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+  factory :target do
+    title { Faker::Lorem.sentence }
+    radius { Faker::Number.decimal(2, 3) }
+    lat { Faker::Number.decimal(2, 3) }
+    lon { Faker::Number.decimal(2, 3) }
```
Comment (martinjaimem) [inline spec/factories/targets_factories.rb:6]: Faker uses only `keyword` arguments, these kinds of arguments are deprecated.

use:
```
Faker::Number.decimal(l_digits: 2, r_digits: 3)
```
instead.


Source: https://github.com/faker-ruby/faker/blob/master/doc/default/number.md

File: spec/requests/targets/targets_create_spec.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  let(:user) { create :user }
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:6]:
```suggestion
  let(:user) { create(:user) }
```

File: spec/requests/targets/targets_create_spec.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  let(:topic) { create :topic }
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:7]:
```suggestion
  let(:topic) { create(:topic) }
```

File: spec/requests/targets/targets_create_spec.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  subject { post v1_targets_path, params: params, as: :json }
+
+  context 'when user is logged in' do
+    sign_in(:user)
+    let(:params) do
+      { target: { title: 'hola', lat: 1, lon: 1, radius: 2, topic_id: topic[:id] } }
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:13]:
```suggestion
      { target: { title: 'Hamburgers', lat: 1, lon: 3, radius: 2, topic_id: topic.id } }
```

Comment (martinjaimem) [reply spec/requests/targets/targets_create_spec.rb:13]: use more "real" data

File: config/routes.rb [hunk 12]
```diff
@@ -6,5 +6,6 @@
       sessions: 'v1/sessions'
     }
     resources :dashboards, only: [:index], as: :dashboards
+    resources :targets, as: :targets
```
Comment (martinjaimem) [inline config/routes.rb:9]: you should test this route

**Comment by martinjaimem** [comment]
You shouldn't have renamed the specs for the user. All specs should end with `_spec.rb`.

**Comment by martinjaimem** [comment]
Add tests 😄

File: db/migrate/20210503145415_create_targets.rb (hunk 1 above)
Comment (ivanfelfo) [reply db/migrate/20210503145415_create_targets.rb:9]: 👍

File: spec/factories/targets_factories.rb (hunk 2 above)
Comment (ivanfelfo) [reply spec/factories/targets_factories.rb:8]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 7 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:5]: 👍

File: spec/factories/targets_factories.rb (hunk 8 above)
Comment (ivanfelfo) [reply spec/factories/targets_factories.rb:6]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 9 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:6]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 10 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:7]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 11 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:13]: 👍

File: spec/requests/targets/targets_create_spec.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:4]:
```suggestion
```

File: spec/requests/targets/targets_create_spec.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+describe 'POST v1/targets', type: :request do
+  let(:user) { create(:user) }
+  let(:topic) { create(:topic) }
+  subject { post v1_targets_path, params: params, as: :json }
+
+  context 'when user is logged in' do
+    sign_in(:user)
+    let(:params) do
+      { target: { title: 'Burgers', latitude: 1, longitude: 1, radius: 2, topic_id: topic[:id] } }
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:13]:
```suggestion
      { target: { title: 'Burgers', latitude: 1, longitude: 1, radius: 2, topic_id: topic.id } }
```

File: spec/requests/targets/targets_create_spec.rb (hunk 3 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:16]: 👍

File: app/models/target.rb (hunk 5 above)
Comment (ivanfelfo) [reply app/models/target.rb:1]: 👍

File: app/models/topic.rb (hunk 6 above)
Comment (ivanfelfo) [reply app/models/topic.rb:1]: 👍

File: config/routes.rb (hunk 12 above)
Comment (ivanfelfo) [reply config/routes.rb:9]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 13 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:4]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 14 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:13]: 👍

File: spec/factories/targets_factories.rb [hunk 15]
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+  factory :target do
+    title { Faker::Lorem.sentence }
+    radius { Faker::Number.decimal(2, 3) }
```
Comment (martinjaimem) [inline spec/factories/targets_factories.rb:4]: same here @ivanfelfo

File: spec/routing/targets_routing_spec.rb [hunk 16]
```diff
@@ -0,0 +1,5 @@
+RSpec.describe 'routes for Targets', type: :routing do
```
Comment (martinjaimem) [inline spec/routing/targets_routing_spec.rb:1]:
```suggestion
describe 'routes for Targets', type: :routing do
```

File: spec/requests/targets/targets_create_spec.rb [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+  end
+  subject { post v1_targets_path, params: params, as: :json }
+
+  context 'when user is logged in' do
+    sign_in(:user)
+
+    it 'creates a target in the db' do
+      subject
+      target = Target.last
+      expect(target.title).to eq('Burgers')
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:15]: what about the latitude/longitude/radius and topic?

File: app/controllers/v1/targets_controller.rb [hunk 18]
```diff
@@ -0,0 +1,23 @@
+module V1
+  class TargetsController < ApplicationController
+    before_action :authenticate_v1_user!
+
+    def create
+      @target = current_v1_user.targets.create!(target_params)
+    end
+
+    def destroy
```
Comment (martinjaimem) [inline app/controllers/v1/targets_controller.rb:9]: please, add tests for it 😄

File: app/models/target.rb (hunk 5 above)
Comment (martinjaimem) [reply app/models/target.rb:1]: I think you can test this using https://github.com/thoughtbot/shoulda-matchers

File: spec/requests/targets/targets_create_spec.rb [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+      subject
+      expect(response).to have_http_status(204)
+    end
+  end
+
+  context 'when user is not logged in' do
+    it 'returns http code 404' do
+      subject
+      expect(response).to have_http_status(401)
+    end
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:32]: add a test to check that the count didn't change

File: spec/models/target_spec.rb [hunk 20]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+RSpec.describe Target, type: :model do
```
Comment (martinjaimem) [inline spec/models/target_spec.rb:3]:
```suggestion
describe Target, type: :model do
```

File: spec/factories/targets_factories.rb (hunk 15 above)
Comment (ivanfelfo) [reply spec/factories/targets_factories.rb:4]: 👍

File: spec/routing/targets_routing_spec.rb (hunk 16 above)
Comment (ivanfelfo) [reply spec/routing/targets_routing_spec.rb:1]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 17 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:15]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 19 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:32]: 👍

File: app/controllers/v1/targets_controller.rb (hunk 18 above)
Comment (ivanfelfo) [reply app/controllers/v1/targets_controller.rb:9]: 👍

File: spec/models/target_spec.rb (hunk 20 above)
Comment (ivanfelfo) [reply spec/models/target_spec.rb:3]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 19 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:36]: 👍

File: spec/models/target_spec.rb (hunk 20 above)
Comment (ivanfelfo) [reply spec/models/target_spec.rb]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 17 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:15]: 👍

File: spec/requests/targets/targets_create_spec.rb [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+    it 'returns http success' do
+      subject
+      expect(response).to have_http_status(204)
+    end
+  end
+
+  context 'when user is not logged in' do
+    it 'returns http code 404' do
```
Comment (juanmferreira93) [inline spec/requests/targets/targets_create_spec.rb:33]: maybe 401 here

File: spec/requests/targets/targets_destroy_spec.rb [hunk 22]
```diff
@@ -0,0 +1,28 @@
+describe 'DELETE /v1/targets/{id}', type: :request do
+  context 'when the user is logged in' do
+    let(:user) { create(:user) }
+    let(:target_of_the_user) { create(:target, user_id: user.id) }
+    let(:other_target) { create(:target) }
+    sign_in(:user)
+
+    it 'deletes a created target that is owned' do
```
Comment (juanmferreira93) [inline spec/requests/targets/targets_destroy_spec.rb:8]: here you can add something like `expect { subject }.not_to change(Target, :count)` in a new `it`

**Review by juanmferreira93 — APPROVED** [review]

File: spec/requests/targets/targets_create_spec.rb (hunk 21 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:33]: 👍

File: spec/requests/targets/targets_destroy_spec.rb (hunk 22 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_destroy_spec.rb:8]: 👍

File: spec/requests/targets/targets_destroy_spec.rb [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+
+  context 'when the user is not logged in' do
+    let(:target) { create(:target) }
+    
+    it 'responds with unauthorized' do
+      delete v1_target_path(target.id)
+      expect(response).to have_http_status(:unauthorized)
+    end
+
+    it 'will NOT change the target count' do
```
Comment (martinjaimem) [inline spec/requests/targets/targets_destroy_spec.rb:29]:
```suggestion
    it 'doesn\'t change the target count' do
```

File: spec/requests/targets/targets_destroy_spec.rb [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+    let(:user) { create(:user) }
+    let(:target_of_the_user) { create(:target, user_id: user.id) }
+    let(:other_target) { create(:target) }
+    sign_in(:user)
+
+    it 'deletes a created target that is owned' do
+      delete v1_target_path(target_of_the_user.id)
+      expect(response).to have_http_status(:success)
+      target_deleted = Target.where(id: target_of_the_user.id).first
+      expect(target_deleted).to be(nil)
```
Comment (martinjaimem) [inline spec/requests/targets/targets_destroy_spec.rb:12]:
```suggestion
      expect(target_of_the_user.reload).to be(nil)
```

File: spec/requests/targets/targets_create_spec.rb [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+    end
+  end
+
+  context 'when user is not logged in' do
+    it 'returns http code 401' do
+      subject
+      expect(response).to have_http_status(401)
+    end
+
+    it 'will NOT change the target count' do
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:38]: same here

File: spec/requests/targets/targets_create_spec.rb [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+    end
+
+    it 'returns http success' do
+      subject
+      expect(response).to have_http_status(204)
+    end
+  end
+
+  context 'when user is not logged in' do
+    it 'returns http code 401' do
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:33]:
```suggestion
    it 'returns http unauthorized' do
```
be consistent

File: app/models/target.rb [hunk 27]
```diff
@@ -0,0 +1,4 @@
...
+  belongs_to :topic
+  belongs_to :user
```
Comment (martinjaimem) [inline app/models/target.rb:3]: add validations of the attributes 🙏

File: app/models/topic.rb [hunk 28]
```diff
@@ -0,0 +1,3 @@
...
+  has_many :targets, dependent: :destroy
```
Comment (martinjaimem) [inline app/models/topic.rb:2]: add validation of presence to the description

File: spec/requests/targets/targets_destroy_spec.rb (hunk 23 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_destroy_spec.rb:29]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 25 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:38]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 26 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:33]: 👍

File: app/models/topic.rb (hunk 28 above)
Comment (ivanfelfo) [reply app/models/topic.rb:2]: topic doesn't have a description. Should I add that field?

File: spec/requests/targets/targets_destroy_spec.rb (hunk 24 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_destroy_spec.rb:12]: 👍

File: app/models/topic.rb (hunk 28 above)
Comment (martinjaimem) [reply app/models/topic.rb:2]: I meant the name

File: spec/requests/targets/targets_destroy_spec.rb [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+  context 'when the user is logged in' do
+    let(:user) { create(:user) }
+    let!(:target_of_the_user) { create(:target, user_id: user.id) }
+    let(:other_target) { create(:target) }
+    sign_in(:user)
+
+    it 'deletes a created target that is owned' do
+      delete v1_target_path(target_of_the_user.id)
+      expect(response).to have_http_status(:success)
+      # target_deleted = Target.where(id: target_of_the_user.id).first
```
Comment (martinjaimem) [inline spec/requests/targets/targets_destroy_spec.rb:11]: 🤔

File: spec/requests/targets/targets_destroy_spec.rb [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+      expect(target_of_the_user.persisted?).to be(true)
```
Comment (martinjaimem) [inline spec/requests/targets/targets_destroy_spec.rb:12]:
```suggestion
```

File: spec/requests/targets/targets_destroy_spec.rb (hunk 29 above)
Comment (martinjaimem) [inline spec/requests/targets/targets_destroy_spec.rb:11]:
```suggestion
      target_deleted = Target.find_by(id: target_of_the_user.id)
```

Comment (ivanfelfo) [reply spec/requests/targets/targets_destroy_spec.rb]: 👍

File: spec/requests/targets/targets_destroy_spec.rb (hunk 30 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_destroy_spec.rb]: 👍

File: app/models/topic.rb (hunk 28 above)
Comment (ivanfelfo) [reply app/models/topic.rb:2]: 👍

File: app/models/target.rb (hunk 27 above)
Comment (ivanfelfo) [reply app/models/target.rb:3]: 👍

File: spec/routing/targets_routing_spec.rb [hunk 31]
```diff
@@ -0,0 +1,5 @@
+describe 'routes for Targets', type: :routing do
+  it 'routes POST /v1/targets to the CREATE action' do
+    expect(post('/v1/targets')).to route_to('v1/targets#create')
+  end
```
Comment (martinjaimem) [inline spec/routing/targets_routing_spec.rb:4]: destroy is missing

File: spec/requests/targets/targets_create_spec.rb [hunk 32]
```diff
@@ -0,0 +1,42 @@
...
+
+    it 'will change the target count by 1' do
+      expect { subject }.to change(Target, :count).by(1)
+    end
+
+    it 'returns http success' do
+      subject
+      expect(response).to have_http_status(200)
+    end
+  end
```
Comment (martinjaimem) [inline spec/requests/targets/targets_create_spec.rb:30]: what happens if the readius is less than 0?

**Review by martinjaimem — APPROVED** [review]

File: spec/routing/targets_routing_spec.rb (hunk 31 above)
Comment (ivanfelfo) [reply spec/routing/targets_routing_spec.rb:4]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 32 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:41]: 👍

**Pull request merged**